### PR TITLE
[Basic] Add a minimal OrderedDictionary data structure

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
<rdar://problem/43693187> Add OrderedDictionary data structure to SwiftPM

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@swift-ci smoke test

File: Sources/Basic/OrderedDictionary.swift [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    ///
+    /// Use this method instead of key-based subscripting when you need to know
+    /// whether the new value supplants the value of an existing key. If the
+    /// value of an existing key is updated, `updateValue(_:forKey:)` returns
+    /// the original value.
+    @discardableResult
+    public mutating func updateValue(_ value: Value, forKey key: Key) -> Value? {
+        // If there is already a value for this key, replace and return the old value.
+        if let oldValue = dict[key] {
+            dict[key] = value
```
Comment (jakepetroules) [inline Sources/Basic/OrderedDictionary.swift]: need to update the array here?

File: Tests/BasicTests/OrderedDictionaryTests.swift [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        XCTAssertEqual(dict.description, "[a: aa, b: bb, c: cc, d: dd]")
+
+        dict["a"] = "aaa"
+        XCTAssertEqual(dict.description, "[a: aaa, b: bb, c: cc, d: dd]")
+
+        dict["e"] = "ee"
+        XCTAssertEqual(dict.description, "[a: aaa, b: bb, c: cc, d: dd, e: ee]")
+
+        dict["b"] = nil
+        XCTAssertEqual(dict.description, "[a: aaa, c: cc, d: dd, e: ee]")
```
Comment (jakepetroules) [inline Tests/BasicTests/OrderedDictionaryTests.swift]: need a test for replacing existing key

File: Sources/Basic/OrderedDictionary.swift (hunk 1 above)
Comment (aciidb0mb3r) [reply Sources/Basic/OrderedDictionary.swift]: no, the position of key remains the same.

File: Sources/Basic/OrderedDictionary.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    /// The element type of a dictionary: a tuple containing an individual
+    /// key-value pair.
+    public typealias Element = (key: Key, value: Value)
+
+    /// The underlying storage for the OrderedDictionary.
+    fileprivate var array: [Key]
+    fileprivate var dict: [Key: Value]
+
+    /// Create an empty OrderedDictionary object.
+    public init() {
```
Comment (jakepetroules) [inline Sources/Basic/OrderedDictionary.swift]: A dictionaryLiteral initializer would be nice

Comment (aciidb0mb3r) [reply Sources/Basic/OrderedDictionary.swift]: Yep, there is an extension for that.

File: Tests/BasicTests/OrderedDictionaryTests.swift (hunk 2 above)
Comment (aciidb0mb3r) [reply Tests/BasicTests/OrderedDictionaryTests.swift]: there is one

**Pull request merged**